tsl/src/continuous_aggs/invalidation_threshold.c

```diff
@@ -56,149 +56,235 @@
+ * to ensure we do not lose invalidations; all transactions that read the
+ * invalidation threshold must have either completed, or not yet read the
+ * value. If a weaker lock was used, an insert transaction that is concurrent
+ * with a refresh could race to read the old threshold before it was updated
+ * and invalidations written during the insert (but after the update) would be
+ * lost.
+ *
+ * Since the invalidation threshold is currently stored in a shared table, all
+ * INSERTs (that read the threshold) and refreshes of continuous aggregates
+ * will be blocked. Therefore, we try to avoid heavy locking if possible by
+ * first reading the threshold with a read lock. If the threshold needs no
+ * update, we need not grab any further locks.
+ */
+bool
+continuous_agg_invalidation_threshold_set(int32 hyper_id, int64 new_threshold)
 {
-	int64 threshold = 0;
-	ScanKeyData scankey[1];
+	Catalog *catalog = ts_catalog_get();
+	Oid threshold_relid = catalog_get_table_id(catalog, CONTINUOUS_AGGS_INVALIDATION_THRESHOLD);
+	Relation rel;
+	ItemPointerData tid;
+	int64 threshold;
+	bool threshold_written = false;
 
-	ScanKeyInit(&scankey[0],
-				Anum_continuous_aggs_invalidation_threshold_pkey_hypertable_id,
-				BTEqualStrategyNumber,
-				F_INT4EQ,
-				Int32GetDatum(hypertable_id));
-
-	if (!ts_catalog_scan_one(CONTINUOUS_AGGS_INVALIDATION_THRESHOLD /*=table*/,
-							 CONTINUOUS_AGGS_INVALIDATION_THRESHOLD_PKEY /*=indexid*/,
-							 scankey /*=scankey*/,
-							 1 /*=num_keys*/,
-							 invalidation_threshold_tuple_found /*=tuple_found*/,
-							 AccessShareLock /*=lockmode*/,
-							 CONTINUOUS_AGGS_INVALIDATION_THRESHOLD_TABLE_NAME /*=table_name*/,
-							 &threshold /*=data*/))
-		elog(ERROR, "could not find invalidation threshold for hypertable %d", hypertable_id);
+	MemSet(&tid, 0, sizeof(tid));
+	Assert(!ItemPointerIsValid(&tid));
+	threshold = invalidation_threshold_read(hyper_id, &tid, AccessShareLock);
 
-	return threshold;
+	if (ItemPointerIsValid(&tid))
+	{
+		/* Found existing threshold. Check if update is needed, otherwise
+		 * return. */
+		if (new_threshold <= threshold)
+		{
+			elog(DEBUG2, "invalidation threshold for hypertable %d needs no update", hyper_id);
+			return false;
+		}
+
+		rel = table_open(threshold_relid, AccessExclusiveLock);
+
+		/* Fetch the latest threshold in case someone updated it
+		 * before we got exclusive lock. */
+		threshold = invalidation_threshold_fetch(rel, &tid);
+
+		if (new_threshold > threshold)
+		{
+			elog(DEBUG2,
+				 "updating existing invalidation threshold from " INT64_FORMAT " to " INT64_FORMAT,
+				 threshold,
+				 new_threshold);
+			invalidation_threshold_write(rel, hyper_id, new_threshold, &tid);
+			threshold_written = true;
+		}
+		else
+		{
+			elog(DEBUG2,
+				 "tried to set invalidation threshold to " INT64_FORMAT
+				 " but it was updated to " INT64_FORMAT " before getting update lock, skipping",
+				 new_threshold,
+				 threshold);
+		}
+	}
+	else
+	{
```

Comment:
I am wary of adding the optimizations from the last commit  (avoiding accessexclusivelock unless necessary) without the other pieces of the code in place.   I don't  understand the whole locking picture (that will come with the other PRs)  and that's my reasoning behind the caution here.

Reply:
I reverted the optimizations and will submit this as a separate PR.